core/src/services/compfs/reader.rs

```diff
@@ -47,14 +47,18 @@ impl oio::Read for CompfsReader {
     async fn read(&mut self) -> Result<Buffer> {
         let pos = self.offset;
         if let Some(end) = self.end {
-            if end >= pos {
+            if end <= pos {
                 return Ok(Buffer::new());
             }
         }
 
         let mut bs = self.core.buf_pool.get();
         // reserve 64KB buffer by default, we should allow user to configure this or make it adaptive.
-        bs.reserve(64 * 1024);
+        if let Some(end) = self.end {
+            bs.reserve((end - pos) as usize);
+        } else {
+            bs.reserve(64 * 1024);
```

Comment:
Another issue might be here. If the file size is `1MiB` and we are reading `1024..2048`. We might read too much data than expected.

If `end` is known, we should set the correct `bs` size.

Reply:
Yes, if `end` is known, the correct `bs` size is set just as you have pointed out. Line 58.

Reply:
`reserve` doesn't guarantee that the buffer size will be exactly `size`; it may end up being larger than expected. Maybe we should call `truncate` to make sure about it?

Reply:
Well, maybe it's better to slice it.

Reply:
Fixed. I think it should work well now.